### ISSUE TITLE
Validation of secrets

### ISSUE THREAD

**Issue by CameronLonsdale** [praetorian-inc/noseyparker#65]
I'd love to use Nosey Parker for defensive purposes, and being able to filter results to only show secrets which are confirmed valid is essential.

It should be an optional feature, likely occurring during the `report` command to keep scans performant.

Each rule could have optional section listing URL + parameters / headers to specify what network request to make in order to verify the credential. The slack rule https://github.com/praetorian-inc/noseyparker/blob/74098e8881bd68a6368032b78b56ddad5ecbbad4/crates/noseyparker/data/default/rules/slack.yml could have a section which specifies `curl https://slack.com/api/auth.test -H "Authorization: Bearer $(1)"` as the request to make. With $(1) as a placeholder for the first capture group from the regex.

Many project re-invent their own request / response matching format, very curious if you might be able to integrate an existing system like https://hurl.dev with a flexible way to test a response to confirm a valid secret.

**Comment by bradlarsen** [comment]
Yes, automated or semi-automated secret validation would be very useful.

This does not exist presently, not due to lack of interest, but for focus and engineering constraints.

The design space for validation mechanisms is large, so thank you for the implementation sketch and pointing at some existing resources. Ideally, a useful validation mechanism could be specified alongside the rule YAML data files (and not as special-cased code in Nosey Parker). I will ruminate on this more.

Thank you for writing this up!

**Comment by bradlarsen** [comment]
Slightly related is #59, a proposal for another _invalidation_ mechanism for findings:

There are several existing rules, like for detecting JWTs and PEM-encoded private keys, where some additional downstream parsing validation could eliminate lots of false positives, but it's not possible to express using regular expressions. That is, if a supposed PEM-encoded private key doesn't actually decode as PEM, ignore it.

**Comment by CameronLonsdale** [comment]
Yeah interesting, another example of invalidation I can think of would be the Atlassian API token, since there is a CRC checksum at the end which would help rule out false positives. But at that point, you may aswell send a request to validate the token and skip the middle step. https://github.com/trufflesecurity/trufflehog/issues/649#issuecomment-1537564184

Design space is very large, agree that having it as part of the YAML rather than special case code would be best as it allows more contributors and customization, versus the way Trufflehog is implemented.

**Comment by munntjlx** [comment]
Also interesting are things like Password: ${{secret}} I get lots of these and they are easily regexable. IN this case it means I look up the secret value from a vault and replace the ${secret} or whatnot with the actual secret but NOT in the code.
